packages/lisk-core/src/controller.js

```diff
@@ -1,40 +1,171 @@
 
 module.exports = class Controller {
 	constructor(options) {
-		console.log('Initializing controller');
+		logger.info('Initializing controller');
+		this.config = options;
 		this.modules = {};
+		this.channel = null;
 
 		// Setting up bus
-		if(!this.bus) {
+		if (!this.bus) {
 			this.bus = new Bus(this, {
 				wildcard: true,
 				delimiter: ':',
-				maxListeners: 1000
+				maxListeners: 1000,
 			});
 		}
 	}
 
-	async loadModules() {
-		const chainModule = ModuleFactory.create('child_process', 'lisk-chain', {}, this, this.bus);
-		const p2p2Module = ModuleFactory.create('in_memory', 'lisk-p2p', {}, this, this.bus);
+	async setup() {
+		await Schema.setup();
+
+		const errors = Schema.validate(this.config, Schema.getSchema().config);
+		if (errors.length) {
+			logger.error(errors);
+			throw new Error('Configuration Validation Failed');
+		}
 
-		this.modules[chainModule.alias] = chainModule;
-		this.modules[p2p2Module.alias] = p2p2Module;
+		process.title = `Lisk ${this.config.pkg.version} : (${
+			this.config.dirs.root
+		})`;
 
-		await Promise.map(Object.keys(this.modules), async (m) => {
-			await this.modules[m].load();
-		});
+		// Make sure all directories exists
+		fs.emptyDirSync(this.config.dirs.temp);
+		Object.keys(this.config.dirs).forEach(dir =>
+			fs.ensureDirSync(this.config.dirs[dir]),
+		);
+		fs.writeFileSync(`${this.config.dirs.pids}/controller.pid`, process.pid);
+
+		// Set the custom directory to load modules
+		if (this.config.dirs.modules) {
+			logger.info(
+				`Setting custom module directory to: ${this.config.dirs.modules}`,
+			);
+		}
 
-		console.info(`Bus listening to events: ${this.bus.getEvents()}`);
-		console.info(`Bus ready for actions: ${this.bus.getActions()}`);
+		// Setup bus RPC socket
+		await this.bus.setup();
+	}
+
+	async load() {
+		await this.establishChannel();
+		await this.loadModules();
+
+		logger.info('Bus listening to events', this.bus.getEvents());
+		logger.info('Bus ready for actions', this.bus.getActions());
+
+		this.channel.publish('lisk:ready', {});
 	}
 
 	getModule(alias) {
 		return this.modules[alias];
 	}
+
+	async establishChannel() {
+		this.channel = new EventEmitterChannel(
+			'lisk',
+			['ready'],
+			['getComponentConfig'],
+			this.bus,
+			{ skipInternalEvents: true },
+		);
+
+		await this.channel.registerToBus();
+
+		this.channel.action(
+			'getComponentConfig',
+			async action => this.config.components[action.params],
+		);
+
+		// If log level is greater than info
+		if (logger.level() < 30) {
+			this.bus.onAny((name, event) => {
+				logger.debug(
+					`MONITOR: ${event.source} -> ${event.module}:${event.name}`,
+					event.data,
+				);
+			});
+		}
+	}
+
+	async loadModules() {
+		this.config.modules.forEach(moduleConfig => {
+			const npmPackage = this.config.dirs.modules
+				? `${this.config.dirs.modules}/${moduleConfig.npmPackage}`
+				: moduleConfig.npmPackage;
+			const module = ModuleFactory.create(
```

Comment:
Why we need ModuleFactory since we can simply check if `this.modules` has the module alias before creating the module?